Ha fixing datasheet and datasheetSpatRaster bugs

hfadams opened the pull request:
Updated datasheet and datasheetSpatRaster functions to:

* Fail with an informative warning message if `filterColumn` or `filterValue` does not exist in the datasheet
* Accept vectors as the `filterValue` argument to filter on multiple values at once (integers and strings)

Updated datasheet function to:
* Handle multi-scenario and multi-project arguments without duplicated rows or errors
* Handle filtering on multi-scenario and multi-project arguments
* Return datasheet without error for single scenarios when `returnScenarioInfo = TRUE`

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Documentation**
  * Clarified filterValue accepts scalar or vector; documented project argument as a single numeric Project ID and tidied examples/formatting.

* **New Features**
  * Multi-signature raster datasheet support with deterministic raster assembly, optional path-only returns, and multi-scenario aggregation.

* **Bug Fixes & Improvements**
  * Stronger input validation, clearer errors/warnings, improved ID resolution and path fallback, and consistent return shapes.

* **Refactor**
  * Unified and centralized public datasheet interfaces and dispatching.

* **Chore**
  * Bumped package RoxygenNote metadata.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

hfadams committed:
adding error message for filterValue and filterColumn not present in datasheet

hfadams committed:
updating datasheet function to accept a vector for filterValue argument

hfadams committed:
updating datasheet and datasheetSpatRaster to accept filterValue vector

hfadams committed:
updating documentation for datasheet and datasheetSpatRaster

hfadams committed:
removed unnecessary comments

hfadams committed:
renamed ID to Id in column names

hfadams committed:
fixed bug with vector of filterValue strings

hfadams committed:
fixed duplicated rows in multi-scenario call

hfadams committed:
fixed merge when returnScenarioInfo = TRUE

hfadams committed:
fixed filtering for multiple scenarios

hfadams committed:
fixed filtering for multiple projects

hfadams committed:
removed unused newColID object

hfadams commented on R/datasheet.R:
```diff
@@ -456,18 +487,16 @@ setMethod("datasheet",
-        } else {
-          filterColumn <- paste0(filterColumn, "=", filterValue)
+
         }
```
I removed `newColID` assignment since it is not used later in the script

katieb1 commented on R/datasheet.R:
```diff
@@ -584,23 +613,79 @@ setMethod("datasheet",
+
+          filteringDone <- FALSE
+
+          if (!is.null(filterColumn) && length(sid) == 1 && length(pid) == 1) { # console filtering for single scenarios or projects only
```
Just make sure that your code doesn't exceed the 80 character limit per line and that comments are consistently formatted (i.e., above code rather than inline)

hfadams replied:
I updated all of my comments to be less than 80 characters long and removed inline comments. I did some reformatting of the script but did not want to change it too much, let me know if you would like me to reformat all lines longer than 80 characters.

katieb1 commented on R/datasheet.R:
```diff
@@ -504,6 +528,11 @@ setMethod("datasheet",
       # => These send you to query building (case for BOTH fastQuery and UseConsole are FALSE) if :
       # sql statement is complex, or more than one proj/sce is provided
+
+      # Disable console filtering for multiple scenarios - we'll filter after merge
```
On line 524 there is the statement:
```
    # Use console if filterColumn argument is used
    if (!is.null(filterColumn)){
      useConsole <- TRUE
      }
```
Can you just update line 524 instead of lines 533-535 to improve readability? Something like:
```
    if (!is.null(filterColumn)){
      if ((length(sid) > 1 || length(pid) > 1) {
        useConsole <- FALSE
      } else {
        useConsole <- TRUE
    }}
```

hfadams replied:
I've edited this section to improve readability

katieb1 commented on R/datasheet.R:
```diff
@@ -815,7 +940,7 @@ setMethod("datasheet",
             if (nrow(lookupSheet) > 0) {
               lookupSheet <- lookupSheet[order(lookupSheet[[names(lookupSheet[1])]]), ]
-              lookupLevels <- lookupSheet[[displayMem]]
+              lookupLevels <- unique(lookupSheet[[displayMem]])
```
There should only ever be one unique "displayMem". I think this was probably caused by that issue with the capitalized ID in ProjectID that we fixed on line 920.

hfadams replied:
I have removed the `unique()` wrapper from the `lookupLevels` assignment, I had added it while debugging an earlier issue with the multi-project function call but it is no longer needed.

katieb1 commented on R/datasheet.R:
```diff
@@ -456,18 +487,11 @@ setMethod("datasheet",
-          filterColumn <- paste0(filterColumn, "=", newColID)
-        } else {
-          filterColumn <- paste0(filterColumn, "=", filterValue)
+          matchedRows <- inputDatasheet$Name %in% filterValue
```
Does this `matchedRows` variable ever get used again?

hfadams replied:
Not in this pathway, it is reassigned at the multi-scenario filtering step. I have removed it here.

katieb1 commented on R/datasheet.R:
```diff
@@ -669,6 +754,45 @@ setMethod("datasheet",
+    }
     
     # TODO review this, this bit assign the correct data types 
     if (empty | lookupsAsFactors | !returnInvisible) {
```
When filter columns are used, I don't think this conditional needs to be entered since you are exporting the validation sheets that contain the pre-converted strings. You can set `lookupsAsFactors <- FALSE` inside the conditionals for when filter values are used (i.e., right inside `if (!is.null(filterColumn) && length(sid) == 1 && length(pid) == 1) ` and `if (!is.null(filterColumn) && (length(sid) > 1 || length(pid) > 1) && exists("originalFilterColumn"))`. This will speed up processing time quite a bit.

hfadams replied:
I set `lookupsAsFactors <- FALSE` inside the conditional for single scenario functions calls `(if (!is.null(filterColumn) && length(sid) == 1 && length(pid) == 1) )` to reduce processing time. I did not add it to the conditional for multi-scenario function calls since it is necessary for returning stratum ID and transition type ID strings

katieb1 commented on R/datasheetSpatRaster.R:
```diff
@@ -217,12 +217,12 @@ setMethod("datasheetSpatRaster", signature(ssimObject = "Scenario"),
   
   # TO DO: make sure datasheet is spatial after opening
-  cMeta <- .datasheet(x, name = datasheet, optional = TRUE, 
+  cMeta <- datasheet(x, name = datasheet, optional = TRUE, 
```
I think these got converted back from the internal wrapper functions

hfadams replied:
I've re-added the internal wrappers

katieb1 reviewed (changes requested):
All is working now! I've just added some review comments for clarity, robustness, and efficiency

hfadams committed:
edited and formatted comments

hfadams committed:
reformatted scripts to reduce characters per line

hfadams committed:
updated logic for selecting filtering pathway

hfadams committed:
re-adding internal wrappers

hfadams committed:
removed unused assignment of matchedRows

hfadams committed:
removing "unique" wrapper from lookupLevels assignment

hfadams committed:
set  lookupsAsFactors <- FALSE for single scenario filtering

hfadams committed:
added guard against empty filterValue

hfadams commented on R/datasheet.R:
```diff
@@ -425,24 +423,56 @@ setMethod("datasheet",
         
         if (is.null(filterValue)) {
           stop("filterColumn specified without a filterValue.")
         }
```
I have added a check for empty `filterValue` inputs

hfadams commented on R/datasheet.R:
```diff
@@ -425,24 +423,56 @@ setMethod("datasheet",
+              paste(missingVals, collapse = ", ")
+            )
+          )
+        }
```
It is expected that if the first scenario contains the specified `filterValue` then the subsequent scenario will as well. For now it is best to have some validation but avoid the more expensive validation of checking all scenarios, so no changes need to be made to this verification step.

hfadams commented on R/datasheetSpatRaster.R:
```diff
@@ -88,325 +88,475 @@ NULL
+    scnList <- .scenario(ssimObject, scenario = scenario)
+    scenario <- NULL
+
+    return(datasheetSpatRaster(
+      scnList,
+      datasheet,
+      column,
+      scenario,
+      iteration,
+      timestep,
+      filterColumn,
+      filterValue,
+      subset,
+      forceElements
+    ))
```
It is not necessary for the `pathOnly` parameter to be passed through the `datasheetSpatRaster` call, the default value can be used.

hfadams committed:
removing duplicated cMeta block

hfadams commented on R/datasheetSpatRaster.R:
```diff
@@ -88,325 +88,475 @@ NULL
+        filterValue = filterValue,
+        lookupsAsFactors = getFactors
+      )
+    }
```
I have removed the duplicated code block for creating cMeta

hfadams commented on R/datasheet.R:
```diff
@@ -9,946 +9,1394 @@ NULL
+    cScn <- ssimObject[[1]]
+    x <- NULL
+
+    if (is(cScn, "Scenario")) {
+      x <- getIdsFromListOfObjects(
+        ssimObject,
+        expecting = "Scenario",
+        scenario = scenario,
+        project = project
+      )
+      scenario <- x$objs
+      project <- NULL
+    }
+
+    if (is(cScn, "Project")) {
+      x <- getIdsFromListOfObjects(
+        ssimObject,
+        expecting = "Project",
+        scenario = scenario,
+        project = project
+      )
+      project <- x$objs
+      scenario <- NULL
+    }
```
ssimObject does not need a check for empty lists. If the user inputs a ssimObject correctly it will not be empty

hfadams commented on R/datasheet.R:
```diff
@@ -9,946 +9,1393 @@ NULL
               if (nrow(lookupSheet) > 0) {
-                if (length(intersect(displayMem, names(lookupSheet))) == 0) {
-                  stop(paste0("Something is wrong. Expecting ", displayMem, 
-                              " in lookup table."))
+                lookupSheet <- lookupSheet[
+                  order(lookupSheet[[names(lookupSheet[1])]]),
+                ]
+                lookupLevels <- lookupSheet[[displayMem]]
+              } else {
+                lookupLevels <- c()
+              }
```
`lookupSheet` is guaranteed to have unique display member values, there is no need for the `unique()` wrapper

hfadams commented on R/datasheet.R:
```diff
@@ -9,946 +9,1394 @@ NULL
+              lib = .filepath(x),
+              sheet = name
+            )
+            tt <- command(args, session = session(x))
+            datasheetCols <- .dataframeFromSSim(tt, csv = FALSE)
+
+            inputDatasheetName <- subset(
+              datasheetCols,
+              name == originalFilterColumn
+            )$formula1
+
+            if (inputDatasheetName != "N/A") {
+              tempFile2 <- tempfile(fileext = ".csv")
+              args <- list(
+                export = NULL,
+                lib = .filepath(x),
+                sheet = inputDatasheetName,
+                file = tempFile2,
+                valsheets = NULL,
+                extfilepaths = NULL,
+                includepk = NULL,
+                force = NULL
+              )
+              args <- assignPidSid(args, sheetNames, pid[1], sid[1])
+              tt <- command(args, session = session(x))
+              inputDatasheet <- read.csv(
+                tempFile2,
+                as.is = TRUE,
+                encoding = "UTF-8"
+              )
+              unlink(tempFile2)
+
+              matchedRows <- inputDatasheet$Name %in% originalFilterValue
+              filterIDs <- inputDatasheet[matchedRows, ][[originalFilterColumn]]
+
+              if (length(filterIDs) > 0) {
+                originalFilterValue <- filterIDs
+              }
+            }
           }
```
As addressed in your earlier comment, this current approach provides verification of the presence of `filterValue` in the first scenario but does not take the more costly approach of verifying that it is present in each scenario. It is expected that if `filterValue` is present in the first scenario it will be present in all scenarios, and the verification for the first scenario is primarily to address typos or otherwise incorrectly input filter values

hfadams commented on R/datasheet.R:
```diff
@@ -9,946 +9,1397 @@ NULL
+    rawValues,
+    verbose
+  ) {
+    cScn <- ssimObject[[1]]
```
If a valid ssimObject has been provided it will not be empty. The user is required to input a valid ssimObject

hfadams committed:
return list in datasheetSpatraster if scenarios are empty

hfadams commented on R/datasheetSpatRaster.R:
```diff
@@ -88,325 +88,475 @@ NULL
+        filterValue = filterValue,
+        subset = subset,
+        forceElements = forceElements,
+        pathOnly = pathOnly
+      )
+
+      if (!(is(cOut, "list") && (length(cOut) == 0))) {
+        names(cOut) <- paste0("scn", .scenarioId(cScn), ".", names(cOut))
+
+        if (!started) {
+          out <- cOut
+        } else {
+          out <- c(out, cOut) #TODO: TEST - changed for terr
+        }
+        started <- TRUE
       }
-      started <- TRUE
     }
+
+    if ((length(names(out)) == 1) & !forceElements) {
+      out <- out[[1]]
+    }
+
+    return(out)
```
I have updated this section to return an empty list if the scenarios are empty (an unlikely edge case) and updated the logical evaluation to use `&&`

hfadams committed:
updating & to &&

hfadams commented on R/datasheetSpatRaster.R:
```diff
@@ -88,325 +88,464 @@ NULL
+        length(intersect(names(cMeta), c("Timestep", "Iteration"))) == 0
+    ) {
+      cMeta$outName <- paste0(cMeta$outName, ".b", cMeta$bandColumn)
     }
```
`cMeta$Band` is populated with NA values, so the column reference does not need to change. I have updated the logical operation to use `&&`

hfadams commented on R/datasheetSpatRaster.R:
```diff
@@ -88,325 +88,464 @@ NULL
-        
-        cRow$rasterColumn <- addPath
+        cMeta$rasterColumn <- addPath
       }
```
`cMeta$Band[1]` is a valid reference, so it does not need to change to `cMeta$bandColumn[1]`. I have updated the logical operation to use `&&`.

hfadams committed:
updating documentation to remove multiple project option

hfadams committed:
added error message for multi-project function calls

hfadams commented on R/datasheetSpatRaster.R:
```diff
@@ -88,325 +88,467 @@ NULL
+    forceElements,
+    pathOnly
+  ) {
+    if (!is(ssimObject[[1]], "Scenario")) {
+      stop(
+        "Expecting an SsimLibrary/Project/Scenario or list of Scenario objects."
+      )
+    }
```
If the user inputs a valid `ssimObject` then `ssimObject[[1]]` will not be empty. Please refer to my earlier comment regarding this.

hfadams commented on R/datasheet.R:
```diff
@@ -9,946 +9,1400 @@ NULL
-                   rawValues, verbose) {
-  return(SyncroSimNotFound(ssimObject))
-})
+setMethod(
+  "datasheet",
+  signature(ssimObject = "character"),
+  function(
+    ssimObject,
+    name,
+    project,
+    scenario,
+    summary,
+    optional,
+    empty,
+    filterColumn,
+    filterValue,
+    lookupsAsFactors,
+    sqlStatement,
+    includeKey,
+    fastQuery,
+    returnScenarioInfo,
+    returnInvisible,
+    rawValues,
+    verbose
+  ) {
+    return(SyncroSimNotFound(ssimObject))
+  }
+)
```
Thank you for noting this, we will address the missing `forceElements` parameter in a future update

hfadams commented on R/datasheetSpatRaster.R:
```diff
@@ -88,325 +88,467 @@ NULL
+        if (!started) {
+          out <- cOut
+        } else {
+          out <- c(out, cOut) #TODO: TEST - changed for terr
```
Thank you for noting this, we will test the terra migration in a future update

hfadams commented on R/datasheet.R:
```diff
@@ -9,946 +9,1400 @@ NULL
+    cScn <- ssimObject[[1]]
+    x <- NULL
+
+    if (is(cScn, "Scenario")) {
+      x <- getIdsFromListOfObjects(
+        ssimObject,
+        expecting = "Scenario",
+        scenario = scenario,
+        project = project
+      )
+      scenario <- x$objs
+      project <- NULL
+    }
+
+    if (is(cScn, "Project")) {
+      x <- getIdsFromListOfObjects(
+        ssimObject,
+        expecting = "Project",
+        scenario = scenario,
+        project = project
+      )
+      project <- x$objs
+      scenario <- NULL
+    }
+
+    ssimObject <- x$ssimObject
+
+    if (is.null(ssimObject)) {
+      stop(
+        "Expecting ssimObject to be an SsimLibrary/Project/Scenario, or a list of Scenarios/Projects."
+      )
+    }
```
`cScn` can only be a scenario or project, this does not need to be updated

hfadams commented on R/datasheet.R:
```diff
@@ -9,946 +9,1400 @@ NULL
+            lib <- x
+          } else {
+            lib <- .ssimLibrary(x)
+          }
+          allScns <- scenario(lib, summary = TRUE)
+          if (!is.element("ParentId", names(allScns))) {
+            warning("Missing ParentId info from scenario(summary=TRUE).")
+            allScns$ParentId <- NA
+          }
+          allScns <- subset(
+            allScns,
+            select = c(ScenarioId, ProjectId, Name, ParentId)
+          )
+          allScns$ParentId <- suppressWarnings(as.numeric(allScns$ParentId))
+          parentNames <- subset(allScns, select = c(ScenarioId, Name))
+          names(parentNames) <- c("ParentId", "ParentName")
+          allScns <- merge(allScns, parentNames, all.x = T)
+
+          allScns <- subset(
+            allScns,
+            select = c(ScenarioId, ProjectId, Name, ParentId, ParentName)
+          )
+
+          names(allScns) <- c(
+            "ScenarioId",
+            "ProjectId",
+            "ScenarioName",
+            "ParentId",
+            "ParentName"
+          )
+          allScns <- allScns[allScns$ScenarioId %in% sid, ]
+
+          if (is.element("ScenarioId", names(sheet))) {
+            # if ScenarioId exists use as merge key
+            sheet <- merge(allScns, sheet, by = "ScenarioId", all.y = TRUE)
+          } else {
+            # if ScenarioId was removed (single scenario case) don't remove it again
+            sheet <- merge(allScns, sheet, all.y = TRUE)
+          }
         }
```
The datasheet is returned with additional columns as intended when `returnScenarioInfo = TRUE` so updating the logic is not needed

hfadams committed:
updated error message for multiple project ID arguments

hfadams committed:
updating error message for multiple project inputs

hfadams commented on R/datasheet.R:
```diff
@@ -9,946 +9,1400 @@ NULL
+      complainIfMissing = TRUE
+    )
+    IDColumns <- c("ScenarioId", "ProjectId")
+    browser()
```
I have removed the browser() debug code

hfadams committed:
fixing filtering for multiple filterValues at the project scope

hfadams commented:
@katieb1 I updated the documentation to state that the project ID is a single value and added an error message if multiple project IDs are supplied (note: this does include cases where multiple scenarios are supplied from different projects).

hfadams committed:
updating multiple project error logic

hfadams commented on R/datasheet.R:
```diff
@@ -9,946 +9,1400 @@ NULL
-    stop("Something is wrong in datasheet().")
-  }
-  
-  if ((summary == TRUE) & !optional) {
-    
-    sumInfo <- subset(sumInfo, select = c("scope", "name", "displayName", "order"))
-    sumInfo[order(sumInfo$order), ]
-    sumInfo$order <- NULL
-    sumInfo <- sumInfo[sumInfo$scope == scopeDS,]
-    
-    return(sumInfo)
-  }
-  
-  # Add data info - only for scenario scope datasheets if sid is defined
-  if (summary == TRUE) {
-    
-    # if no scenario scope sheets, return sumInfo without checking for data
-    scnSheetSum <- sum(sumInfo$scope == "scenario")
-    
-    if (scnSheetSum == 0) {
-      
+
+    # now assume one or more names
+    if (is.null(name) & summary == FALSE) {
+      sumInfo <- .datasheets(x, project[[1]], scenario[[1]])
+      allNames <- sumInfo$name
+    } else if (is.null(name) & !summary == FALSE) {
+      stop("Something is wrong in datasheet().")
+    }
+
+    if ((summary == TRUE) & !optional) {
+      sumInfo <- subset(
+        sumInfo,
+        select = c("scope", "name", "displayName", "order")
+      )
       sumInfo[order(sumInfo$order), ]
       sumInfo$order <- NULL
-      sumInfo <- sumInfo[sumInfo$scope == scopeDS,]
-      
+      sumInfo <- sumInfo[sumInfo$scope == scopeDS, ]
```
The function is working as-is, if we come across cases where these parameters are problematic then we will re-evaluate
